estimator/lwe_dual.py

```diff
@@ -428,7 +424,7 @@ def __call__(
 
         params = params.normalize()
 
-        if params.Xs.is_sparse:
+        if type(params.Xs) is SparseTernary:
```

Comment:
`is_sparse` could cover more than merely `SparseTernary`

Reply:
With the current sparsity cutoff at 1/2, I think the only other sparse noise distributions are:
- `Uniform(x, x)`: a constant value is not a secret so this makes LWE trivially broken, and the code will produce some errors.
- DiscreteGaussian with a very small `stddev`: I believe the current code does not accurately handle such Discrete Gaussians. Namely, experimenting with stddev=0.5 show me that `density` and `support_size` are currently computing under- and overestimates. (I'm happy to create an issue for this if needed!)

The reason I changed it to SparseTernary is to 1) take the number of +1's and -1's into account and 2) remove the calculation using the bounds as these are giving overestimates on the actual support size for the above case of the DiscreteGaussian.

Let me split up this PR into only refactoring ND, and I'll put the SparseTernary changes into a separate one.

Reply:
Sounds good. FWIW I was also thinking about future proving the code :)

Reply:
Hum, I created a branch for just refactoring ND: 
https://github.com/malb/lattice-estimator/compare/main...ludopulles:lattice-estimator:refactor-ND

But it seems there are some weird cases, where SparseTernary is called with `p` a half-integer. Is it okay to not split up the PR, but instead to keep the generic "is_sparse" code with a special case for SparseTernary inside there?

Reply:
What do you mean by: "keep the generic "is_sparse" code with a special case for SparseTernary inside there" ? An attribute that checks "isinstance(SparseTernary)"?

Reply:
Basically this commit: 31957df62dddc4517e1a54f46a3812741c12398d

I meant to revert the code back to "if params.Xs.is_sparse:", and then inside this if-statement add code specific for SparseTernary to arrive at the right estimates, because the previous generic sparse secret code did not take into account the weight and +1/-1 conditions for SparseTernary.